File: cpp/src/Vector3D.cpp
#include "Vector3D.h"

namespace kimhyunha {

Vector3D::Vector3D(double x, double y, double z) : Point3D(x, y, z) {}

double Vector3D::magnitude() const {
    return std::sqrt(x() * x() + y() * y() + z() * z());
}

Vector3D Vector3D::normalized() const {
    const double mag = magnitude();
    if (mag == 0.0) {
        return Vector3D(0.0, 0.0, 0.0);
    }
    return *this / mag;
}

double Vector3D::dot(const Vector3D& other) const {
    return x() * other.x() + y() * other.y() + z() * other.z();
}

Vector3D Vector3D::cross(const Vector3D& other) const {
    return Vector3D(
        y() * other.z() - z() * other.y(),
        z() * other.x() - x() * other.z(),
        x() * other.y() - y() * other.x());
}

double Vector3D::distance_to(const Vector3D& other) const {
    return (*this - other).magnitude();
}

Vector3D Vector3D::operator+(const Vector3D& other) const {
    return Vector3D(x() + other.x(), y() + other.y(), z() + other.z());
}

Vector3D Vector3D::operator-(const Vector3D& other) const {
    return Vector3D(x() - other.x(), y() - other.y(), z() - other.z());
}

Vector3D Vector3D::operator-() const {
    return Vector3D(-x(), -y(), -z());
}

Vector3D Vector3D::operator*(double scalar) const {
    return Vector3D(x() * scalar, y() * scalar, z() * scalar);
}

Vector3D Vector3D::operator/(double scalar) const {
    return Vector3D(x() / scalar, y() / scalar, z() / scalar);
}

Vector3D operator*(double scalar, const Vector3D& v) {
    return v * scalar;
}

}  // namespace kimhyunha

File: python/src/libs/Vector3D.py
from __future__ import annotations

import math
from dataclasses import dataclass

from .Point3D import Point3D


@dataclass(frozen=True, slots=True)
class Vector3D(Point3D):
    def __add__(self, other: Vector3D) -> Vector3D:
        return Vector3D(self.x + other.x, self.y + other.y, self.z + other.z)

    def __sub__(self, other: Vector3D) -> Vector3D:
        return Vector3D(self.x - other.x, self.y - other.y, self.z - other.z)

    def __neg__(self) -> Vector3D:
        return Vector3D(-self.x, -self.y, -self.z)

    def __mul__(self, scalar: float) -> Vector3D:
        return Vector3D(self.x * scalar, self.y * scalar, self.z * scalar)

    __rmul__ = __mul__

    def __truediv__(self, scalar: float) -> Vector3D:
        return Vector3D(self.x / scalar, self.y / scalar, self.z / scalar)

    @property
    def magnitude(self) -> float:
        return math.sqrt(self.x**2 + self.y**2 + self.z**2)

    @property
    def normalized(self) -> Vector3D:
        mag = self.magnitude
        if mag == 0:
            return Vector3D(0, 0, 0)
        return self / mag

    def dot(self, other: Vector3D) -> float:
        return self.x * other.x + self.y * other.y + self.z * other.z

    def cross(self, other: Vector3D) -> Vector3D:
        return Vector3D(
            self.y * other.z - self.z * other.y,
            self.z * other.x - self.x * other.z,
            self.x * other.y - self.y * other.x,
        )

    def distance_to(self, other: Vector3D) -> float:
        return (self - other).magnitude
